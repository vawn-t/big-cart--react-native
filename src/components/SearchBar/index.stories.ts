import type { Meta, StoryObj } from '@storybook/react-native';

import SearchBar from '.';

const meta: Meta<typeof SearchBar> = {
  title: 'Components/SearchBar',
  component: SearchBar,
};

export default meta;

type Story = StoryObj<React.ComponentProps<typeof SearchBar>>;

export const Base: Story = {
  args: {},
};
