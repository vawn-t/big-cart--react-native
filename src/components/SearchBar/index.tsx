import { useCallback, useEffect, useState } from 'react';
import {
  NativeSyntheticEvent,
  TextInputSubmitEditingEventData,
} from 'react-native';
import { RouteProp, useNavigation, useRoute } from '@react-navigation/native';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';

// Components
import { Input } from '@components/common';

// Hooks

// Icons
import { IconName, RootStackParamList } from '@interfaces/index';

// Themes
import { colors } from '@themes/index';

// Constants
import { ScreenNames } from '@constants/navigation';

// Stores
import useStore from '@stores/useStore';

interface IProps {
  currentSearchValue?: string;
}

const SearchBar = ({ currentSearchValue = '' }: IProps) => {
  // Stores
  const addSearchItem = useStore.use.addSearchItem();
  const setCurrentSearchItem = useStore.use.setCurrentSearchItem();

  const [searchValue, setSearchValue] = useState(currentSearchValue);

  const route = useRoute<RouteProp<RootStackParamList>>();

  const navigation =
    useNavigation<
      NativeStackNavigationProp<RootStackParamList, ScreenNames.HomeStack>
    >();

  useEffect(() => {
    if (currentSearchValue) {
      setSearchValue(currentSearchValue);
    }
  }, [currentSearchValue]);

  const handleSearch = useCallback(
    (event: NativeSyntheticEvent<TextInputSubmitEditingEventData>) => {
      const value = event.nativeEvent.text;
      if (value) {
        // Only add to history if value is not empty
        addSearchItem(value);
      }

      setCurrentSearchItem(value);

      if (route.name !== ScreenNames.search) {
        // Navigate from home screen
        navigation.navigate(ScreenNames.search);
      }
    },
    [],
  );

  const handleSearchTextChange = useCallback((text: string) => {
    setSearchValue(text);

    if (!text) {
      setCurrentSearchItem('');
    }
  }, []);

  return (
    <Input
      icon={IconName.Search}
      placeholder="Search keywords.."
      value={searchValue}
      containerStyles={{ backgroundColor: colors.background.dark }}
      onSubmitEditing={handleSearch}
      onChangeText={handleSearchTextChange}
    />
  );
};

export default SearchBar;
