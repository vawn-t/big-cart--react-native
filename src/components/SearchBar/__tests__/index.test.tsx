import '@testing-library/jest-native';

import SearchBar from '..';
import { act, fireEvent, render, screen, waitFor } from '@utils/index';
import { renderHook } from '@testing-library/react-hooks';
import useStore from '@stores/useStore';

jest.mock('@react-native-async-storage/async-storage', () =>
  require('@react-native-async-storage/async-storage/jest/async-storage-mock'),
);

const setup = (overrideProps = {}) => {
  return render(<SearchBar {...overrideProps} />);
};

describe('SearchBar component', () => {
  it('renders correctly', () => {
    const { toJSON } = setup();
    expect(toJSON()).toMatchSnapshot();
  });

  it('render correct currentSearchValue', () => {
    setup({ currentSearchValue: 'test' });

    expect(screen.getByDisplayValue('test')).toBeTruthy();
  });

  it('render should update currentSearchValue in store correctly', () => {
    const { result } = renderHook(() => useStore());

    setup();

    fireEvent.changeText(
      screen.getByPlaceholderText('Search keywords..'),
      'test',
    );

    act(() => {
      result.current.setCurrentSearchItem('test');
    });

    expect(result.current.currentSearchValue).toBe('test');

    fireEvent.changeText(screen.getByDisplayValue('test'), '');

    expect(result.current.currentSearchValue).toBe('');
  });

  it('should handle search correctly', async () => {
    const { result } = renderHook(() => useStore());
    setup();

    const input = screen.getByPlaceholderText('Search keywords..');

    act(() => {
      // fireEvent.changeText(input, 'test search');
      fireEvent(input, 'submitEditing', {
        nativeEvent: { text: 'test search' },
      });
    });

    await waitFor(() => {
      expect(result.current.searchHistory.includes('test search')).toBeTruthy();
    });
  });
});
