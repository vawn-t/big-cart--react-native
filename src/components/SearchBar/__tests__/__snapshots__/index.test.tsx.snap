// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar component renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "#FFFFFF",
                  "borderColor": "#FFFFFF",
                  "borderRadius": 5,
                  "borderWidth": 1,
                  "flexDirection": "row",
                },
                {
                  "backgroundColor": "#F4F5F9",
                },
                null,
              ]
            }
            testID="Input"
          >
            <View
              style={
                {
                  "flex": 1,
                  "paddingHorizontal": 28,
                }
              }
            >
              <RNSVGSvgView
                bbHeight="20"
                bbWidth="20"
                fill="none"
                focusable={false}
                height={20}
                style={
                  [
                    {
                      "backgroundColor": "transparent",
                      "borderWidth": 0,
                    },
                    {
                      "flex": 0,
                      "height": 20,
                      "width": 20,
                    },
                  ]
                }
                width={20}
              >
                <RNSVGGroup
                  fill={null}
                  propList={
                    [
                      "fill",
                    ]
                  }
                >
                  <RNSVGPath
                    d="m18.506 17.328-4.438-4.437a7.879 7.879 0 0 0 1.765-4.973C15.833 3.553 12.281 0 7.917 0 3.552.001 0 3.553 0 7.918s3.552 7.916 7.917 7.916a7.879 7.879 0 0 0 4.973-1.765l4.437 4.438a.831.831 0 0 0 1.179 0 .832.832 0 0 0 0-1.179Zm-10.59-3.16a6.256 6.256 0 0 1-6.25-6.25 6.256 6.256 0 0 1 6.25-6.25 6.256 6.256 0 0 1 6.25 6.25 6.256 6.256 0 0 1-6.25 6.25Z"
                    fill={
                      {
                        "payload": 4287006857,
                        "type": 0,
                      }
                    }
                    propList={
                      [
                        "fill",
                      ]
                    }
                  />
                </RNSVGGroup>
              </RNSVGSvgView>
            </View>
            <TextInput
              autoComplete="off"
              onChangeText={[Function]}
              onSubmitEditing={[Function]}
              placeholder="Search keywords.."
              secureTextEntry={false}
              style={
                [
                  {
                    "flex": 10,
                    "paddingVertical": 20,
                  },
                  undefined,
                ]
              }
              testID="Input-TextInput"
              value=""
            />
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "flex": 1,
                  "opacity": 1,
                  "paddingHorizontal": 28,
                }
              }
            />
          </View>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitleVisible={true}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="test"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      />
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
